### PR TITLE
Fix SyncResultObject performances using a Map (fixes #806)

### LINKED ISSUE #806: Performance issue in SyncResultObject
https://bugzilla.mozilla.org/show_bug.cgi?id=1453961

```
<florian> mais c'est assez facile en lisant le code de voir que c'est O(n^2) et que ça devrait pas l'être
<florian> pour le test de déduplication faut utiliser un set, et pas le regénérer à chaque élément
```

Introduced in https://github.com/Kinto/kinto.js/pull/553/files#diff-d2dbe7779d59d4dd594e92248caba0a5

### PR DESCRIPTION
**Pull request by leplatrem**
fixes #806

### CODE REVIEW

File: src/collection.js [hunk 1]
```diff
@@ -74,13 +74,10 @@ export class SyncResultObject {
...
-      const existing = acc.filter(
-        r => (cur.id && r.id ? cur.id != r.id : true)
-      );
-      return existing.concat(cur);
-    }, []);
-    this[type] = deduplicated;
+    const ids = new Set(entries.map(({ id }) => id));
+    this[type] = this[type]
+      .filter(({ id }) => !id || !ids.has(id))
+      .concat(entries);
```
Comment (glasserc) [inline src/collection.js]: What if we see `result.add("skipped", [{id: 1, name: "a"}, {id: 2, name: "b"}])`?

**Comment by glasserc** [comment]
Thinking about this more, maybe we should be using a Map instead of a Set.

**Comment by leplatrem** [comment]
@glasserc r?

File: src/collection.js [hunk 2]
```diff
@@ -72,15 +72,15 @@ export class SyncResultObject {
...
-      const existing = acc.filter(
-        r => (cur.id && r.id ? cur.id != r.id : true)
-      );
-      return existing.concat(cur);
-    }, []);
-    this[type] = deduplicated;
+    const ids = new Map(entries.map(({ id }) => [id, true]));
+    this[type] = this[type]
+      .filter(({ id }) => !id || !ids.has(id))
+      .concat(entries);
```
Comment (glasserc) [inline src/collection.js]: What if we see `result.add("skipped", [{id: 1, name: "a"}, {id: 1, name: "b"}]`?

File: test/collection_test.js [hunk 3]
```diff
@@ -372,8 +372,8 @@ describe("Collection", () => {
         result.add("skipped", [{ name: "d" }]);
         result.add("skipped", [{ name: "e" }]);
         expect(result.skipped).eql([
-          { id: 2, name: "b" },
           { id: 1, name: "c" },
+          { id: 2, name: "b" },
```
Comment (leplatrem) [inline test/collection_test.js]: That's cheating :)

**Comment by leplatrem** [comment]
I'm OK with your changes :) 
Feel free to merge!

**Pull request merged**